frontend: Fix some accessibility issues related to ARIA audit tool.

vaidap opened the pull request:
Continuation of #5519.
Now not having the settings overlay open gives 1 audit error (`a class="search_icon"`), while having the overlay open gives none in terms of the error regarding having obscured/invisible items being focusable.

`Fix links not being clear from link text.` fixes #5535.

smarx commented:
*Automated message from Dropbox CLA bot*

@vaidap, it looks like you've already signed the Dropbox CLA. Thanks!

vaidap commented:
The second commit fixes #5535.

vaidap commented:
would be good to get #5201 merged for this!

vaidap commented:
@zulipbot claim

timabbott commented:
I merged the first commit, after some small changes to whitespace.

timabbott commented on templates/zerver/navbar.html:
```diff
@@ -12,6 +12,7 @@
     <div class="column-middle-inner">
           <div id="streamlist-toggle" {%if embedded %} style="visibility: hidden"{% endif %}>
             <a href="#" id="streamlist-toggle-button" role="button"><i class="icon-vector-reorder"></i>
+              <span class="ariahide">Streamlist toggle button</span>
```
Should this be using some more formal `ARIA` tag than a new CSS class?  E.g. `aria-label` or something?

Also, these strings, like our other user-facing strings, should be tagged for translation.

vaidap replied:
That's a good point, I had initially gone with this approach (https://stackoverflow.com/questions/32631180/link-contain-no-text-showing-error-in-wave-accessibility-evaluation-tool), as it prevents the anchor from being empty. it does seem from https://www.deque.com/blog/text-links-practices-screen-readers/ that aria-label gains precedence on any attribute so it should be the one we use, but I'm not sure if the empty link text causes a different problem, but it surely triggers some accessibility tools (https://purplehippodesign.co.uk/make-font-awesome-accessible/). @jmbowman do you have any thoughts on this?

timabbott commented on templates/zerver/compose.html:
```diff
@@ -100,7 +100,7 @@
                     <label id="enter-sends-label" class="compose_checkbox_label" for="enter_sends">{{ _('Press Enter to send') }}&nbsp;</label>
                     <input type="checkbox" id="enter_sends" name="enter_sends" value="enter_sends" />
-                    <button type="submit" id="compose-send-button" class="button small send_message" tabindex="150">{{ _('Send') }}</button>
+                    <button type="submit" id="compose-send-button" class="button small send_message" tabindex="0">{{ _('Send') }}</button>
```
Does changing this tabindex actually fix a problem?  I vaguely recall that we intentionally set it super high, and testing, it appears that the high tab index means that hitting tab one too many times with compose open does nothing, rather than going to the users list in the right sidebar (which can be annoying for people using `tab+enter` to send messages, since then hitting `tab+tab+enter` means something totally random happens)

vaidap replied:
it is triggering the [Avoid positive integer values for tabIndex](https://github.com/GoogleChrome/accessibility-developer-tools/wiki/Audit-Rules#ax_focus_03) warning, but it's more of a recommendation & here we are consciously using the tabIndex to improve navigation so i'll just remove the commit.

timabbott commented on zerver/lib/bugdown/__init__.py:
```diff
@@ -146,6 +146,7 @@ def add_a(root, url, link, title=None, desc=None,
         a.set("data-id", data_id)
     img = markdown.util.etree.SubElement(a, "img")
     img.set("src", url)
+    img.set("alt", title)
```
I feel like this is another case where I want us to determine what the correct accessibility thing to do is.  Do we need to set  `title`  and `alt` to be the same thing for both this and emoji?  

@jmbowman, can you help track down a good recommendation for how images should be tagged for accessibility that also want a hover-title behavior?  Since changing the HTML for inline images is a somewhat disruptive change, I want to make sure we do it correctly.

jmbowman replied:
The WebAIM site has some [guidelines](http://webaim.org/articles/gonewild/#title) on this.  Basically, the title attribute should only be used for additional information that isn't required for a11y purposes (and will often be skipped by screen readers).

vaidap replied:
and i'd imagine that the hover-title behaviour would only be able to provide users with the title of the image, which is not as useful for images as it is for emoji codes

vaidap commented:
Removed compose button tabIndex change, added some extra spans for links that need descriptive text, and tagged them for translation, and left comments re:how to go about some issues above.

Also noticed some anchors that do not trigger the audit tool because they have title attributes-- since `title` seems to not be the best for screenreaders, I believe adding the spans as a follow-up would be a good idea there too.

vaidap commented:
`Hide x character from screenreaders in exit divs` makes the same type of fixes as c9bc8031189f5ff85f3075e04127879bb9116390, but for new close buttons/ones i had somehow missed.

zulipbot-test commented:
Heads up @vaidap, we just merged some commits that conflict with the changes your made in this pull request! You can review this repository's [recent commits](https://github.com/zulip/zulip/commits/master) to see where the conflicts occur. Please rebase your feature branch against the `upstream/master` branch and [resolve](http://zulip.readthedocs.io/en/latest/git-guide.html#recover-from-a-git-rebase-failure) your pull request's merge conflicts accordingly.

timabbott commented:
I merged the navbar fix.  

@vaidap I'm not sure the `x`s should be hidden for ARIA; won't that make it really hard to use those exit buttons?

timabbott commented:
For the uploaded images, I think the right answer might be to have `alt=""`.  The title of the image is already included in the message in the form of the URL in the message already; so if you're using a screenreader, the image itself is useless, right?

vaidap commented:
* about the image titles, good point, i had forgotten that the image title always appears as a link. changed it to `""` and updated tests.
* regarding the `x`, i think the desired behaviour would be that the div enclosing the `x`, which triggers the closing of the modal/etc, is the one that would get focus & would have an aria-label that would signify that it is a closing button, instead of having `x` or `multiply` being read out. an alternative would be to have some text next to the x that would be hidden from non-screenreader-using users, but would be recognised by screenreaders (similarly to the 844e5666ff9e386c98fedca2bef979c8ff82e50d change)

timabbott commented:
Closing, since basically every line is a merge conflict at this point. Thanks @vaidap!
